pkg/integrations/v4/dm/emitter.go

```diff
@@ -174,90 +173,74 @@ func (e *emitter) processWithoutRegister(metadata integration.Definition, extraL
 	if err != nil {
-		elog.WithError(err).WithField("output", string(integrationJSON)).Warn("can't parse v4 integration output")
-		return err
+		elog.
+			WithError(err).
+			WithField("integration", dto.Metadata.Name).
+			Errorf("cannot determine agent short name")
+		return
 	}
 
-	return e.process(metadata, extraLabels, entityRewrite, pluginDataV4)
-}
-
-func (e *emitter) process(
-	metadata integration.Definition,
-	extraLabels data.Map,
-	entityRewrite []data.EntityRewrite,
-	integrationData protocol.DataV4) (err error) {
-
-	pluginId := metadata.PluginID(integrationData.Integration.Name)
-	plugin := agent.NewExternalPluginCommon(pluginId, e.agentContext, metadata.Name)
-	labels, extraAnnotations := metadata.LabelsAndExtraAnnotations(extraLabels)
+	pluginId := dto.Metadata.PluginID(dto.IntegrationData.Integration.Name)
+	plugin := agent.NewExternalPluginCommon(pluginId, e.agentContext, dto.Metadata.Name)
+	labels, extraAnnotations := dto.Metadata.LabelsAndExtraAnnotations(dto.ExtraLabels)
 
 	var entities []protocol.Entity
-	datasetsByEntityName := make(map[string]protocol.Dataset, len(integrationData.DataSets))
+	datasetsByEntityName := make(map[string]protocol.Dataset, len(dto.IntegrationData.DataSets))
 	// Collect All entities
-	for i := range integrationData.DataSets {
-		entities = append(entities, integrationData.DataSets[i].Entity)
-		datasetsByEntityName[integrationData.DataSets[i].Entity.Name] = integrationData.DataSets[i]
-	}
-
-	agentShortName, err := e.agentContext.IDLookup().AgentShortEntityName()
-	if err != nil {
-		return wrapError(fmt.Errorf("error renaming entity: %s", err.Error()), len(integrationData.DataSets))
+	for _, ds := range dto.IntegrationData.DataSets {
+		entities = append(entities, ds.Entity)
+		datasetsByEntityName[ds.Entity.Name] = ds
 	}
 
 	var emitErrs []error
-	processEntityDataset := func(dataset protocol.Dataset, entityID entity.ID) {
-		// for dataset.Entity call emitV4DataSet function with entity ID
-
-		dataset.Common.Attributes[nrEntityId] = entityID.String()
-		replaceEntityName(dataset.Entity, entityRewrite, agentShortName)
-
-		emitInventory(
-			&plugin,
-			metadata,
-			integrationData.Integration,
-			entityID,
-			dataset,
-			labels,
-		)
-
-		emitEvent(
-			&plugin,
-			metadata,
-			dataset,
-			labels,
-		)
-
-		dmProcessor := IntegrationProcessor{
-			IntegrationInterval:         metadata.Interval,
-			IntegrationLabels:           labels,
-			IntegrationExtraAnnotations: extraAnnotations,
-		}
-
-		metrics := dmProcessor.ProcessMetrics(dataset.Metrics, dataset.Common, dataset.Entity)
-		if err := e.metricsSender.SendMetricsWithCommonAttributes(dataset.Common, metrics); err != nil {
-
-		}
-	}
-
 	registeredEntities, unregisteredEntitiesWithWait := e.RegisterEntities(entities)
 
 	for entityName, entityID := range registeredEntities {
-		processEntityDataset(datasetsByEntityName[entityName], entityID)
+		func(dataset protocol.Dataset, entityID entity.ID) {
```

Comment:
There is no need in closure, if you make it anonymous and newer reuse. It was needed to call the same thing 2 times with all data inside.

Reply:
this code no longer exists on last children PR